### PR TITLE
Use pure tokenizer (#8)

### PR DESCRIPTION
**Pull request by ariya**

### CODE REVIEW

**Comment by ariya** [comment]
> I'd be ok with never obtaining the ast in the first place, or at least turn that off and allow turning it on via an option.

What should be the option name and the default value? `ast` and `false` as the defaut is OK for you?

And I can update the docs accordingly.

**Comment by thlorenz** [comment]
I'd prefer this behavior:

How about the following behavior?

`ast` option is `false` (=default):
 * Behavior: tokenize the input (don't even try to build the AST)
 * Return value: `ast` is not there (undefined)
 * Exception: only thrown if the tokenizer fails

`ast` option is `true`:
 * Behavior: parse the input (no fallback to tokenization)
 * Return value: `ast` is returned when the input is parsed successfully (syntactically correct)
 * Exception: thrown if the parser fails with message that user should try again with `ast: false`

**Comment by ariya** [comment]
I'll adjust the implementation to match the proposed behavior.

Note that this new behavior (`buildAst` is `false` by default) is a breaking change. I can check and tweak cardinal (e.g. it assumes invalid code will throw an exception), but other downstream users might break.

**Comment by thlorenz** [comment]
> Note that this new behavior (buildAst is false by default) is a breaking change

Yes I know, we'll publish a new major version.

**Comment by ariya** [comment]
> Yes I know, we'll publish a new major version.

If you'd like a transition time (only a minor version update), we can do the following:
* Implement the proposed behavior
* Ensure that the default value for `buildAst` to `true`
* Modify cardinal to use redeyed with `buildAst = false`

That way, cardinal can benefit from this pure tokenizer mode while other downstream users have some time to evaluate and migrate.

Entirely up to you, I can live with or without the transition.

**Comment by thlorenz** [comment]
Let's go with breaking things as I think it's better in general to just use the tokenizer by default.

**Comment by ariya** [comment]
> Let's go with breaking things as I think it's better in general to just use the tokenizer by default.

In that case, is there a strong reason to keep the AST mode?

**Comment by thlorenz** [comment]
Yes, since some people may want to build tools around this that use the AST, so let's not totally kill it.

**Comment by ariya** [comment]
@thlorenz Any review/comment on the latest commit?

File: test/redeyed-result.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ test('redeyed result has esprima ast, tokens, comments and splits and transforme
     , tokens =  ast.tokens
     , comments = ast.comments
 
-    , result = redeyed(code, conf)
+    , result = redeyed(code, conf, { buildAst: true } )
```
Comment (thlorenz) [inline test/redeyed-result.js]: Probably should add a test here with default opts. Asserting that it has no `ast` but everything else.

File: test/redeyed-smoke.js [hunk 2]
```diff
@@ -53,9 +53,9 @@ test('esprima', function (t) {
 
   readdirp({ root: esprimadir, fileFilter: '*.js' })
     .on('data', function (entry) {
-      
+
       var code = fs.readFileSync(entry.fullPath, 'utf-8')
-        , result = redeyed(code, { Keyword: { 'var': '+:-' } }).code
+        , result = redeyed(code, { Keyword: { 'var': '+:-' } }, { buildAst: true }).code
```
Comment (thlorenz) [inline test/redeyed-smoke.js]: not sure why we need `ast` here. Shouldn't surrounding still work as it did without an AST?

**Comment by thlorenz** [comment]
Great work so far, just need to figure out those last tidbits. Thanks.

File: test/redeyed-result.js (hunk 1 above)
Comment (ariya) [inline test/redeyed-result.js]: Good point, I'll add another test.

File: test/redeyed-smoke.js (hunk 2 above)
Comment (ariya) [inline test/redeyed-smoke.js]: Now that you mention it, I think the test should check both cases of `buildAst`.

**Comment by ariya** [comment]
@thlorenz Do the added/tweaked tests look good now?

**Comment by thlorenz** [comment]
LGTM @ariya when you feel this is complete feel free to squash+merge to master and ping me please so I can version and publish.

**Comment by ariya** [comment]
@thlorenz Landed in master. Thank you!

**Comment by thlorenz** [comment]
Thanks published as v1.0.0. Major upgrade due to breaking changes.

**Pull request closed without merging**